distributed/scheduler.py

```diff
@@ -5397,6 +5396,58 @@ async def gather(self, comm=None, keys=None, serializers=None):
         self.log_event("all", {"action": "gather", "count": len(keys)})
         return result
 
+    async def _gather_on_workers(
+        self, transfers: "Mapping[str, Mapping[Hashable, list[str]]]"
+    ) -> "dict[str, set[Hashable]]":
+        """Peer-to-peer copy of keys between workers
+
+        Parameters
+        ----------
+        transfers :
+            {recipient address: {key: [sender address, sender address, ...], key: ...}}
+
+        Returns
+        -------
+        returns:
+            {recipient address: {failed key, failed key, ...}}
+        """
+        results = await asyncio.gather(
+            *(
+                retry_operation(self.rpc(addr=w).gather, who_has=who_has)
+                for w, who_has in transfers.items()
+            )
+        )
+
+        parent: SchedulerState = cast(SchedulerState, self)
+        failed_keys_by_recipient = {}  # {recipient address: {failed keys}}
+
+        for w, result in zip(transfers, results):
+            if result["status"] == "OK":
+                keys_fail = set()
+            elif result["status"] == "missing-data":
+                keys_fail = set(result["keys"])
+                logger.warning("Worker %s failed to acquire keys: %s", w, keys_fail)
+            else:
+                keys_fail = set(transfers[w])
+                logger.warning("Communication failed during replication: %s", result)
+            failed_keys_by_recipient[w] = keys_fail
+            keys_ok = transfers[w].keys() - keys_fail
+            ws: WorkerState = parent._workers_dv.get(w)
+            if ws is None:
+                logger.warning("Worker lost during replication: %s", w)
+                continue
+            for key in keys_ok:
+                ts: TaskState = parent._tasks.get(key)
+                if ts is None or ts._state != "memory":
+                    logger.warning("Key lost during replication: %s", key)
+                    continue
+                if ts not in ws._has_what:
+                    ws._nbytes += ts.get_nbytes()
+                    ws._has_what[ts] = None
+                    ts._who_has.add(ws)
+
+            return failed_keys_by_recipient
```

Comment:
I wonder if this function could be broken down by key or by recipient address.  This seems like the kind of thing that we might be able to do as effectively "in the small" rather than as a bulk operation.  

I suspect that there are some advantages to bulk here when we think about sending multiple keys out from the same worker (using a different RPC call for each key seems unpleasant).  But maybe it makes sense to call one coroutine here per worker rather than one coroutine for all of the workers at once.

This isn't that important.  I mostly suggest this out of principle.  I _think_ (although could be wrong) that treating transfers in a more fine-grained way will help us in the long run.

Pragmatically this would move the for loop from this function into where the function is called.

Reply:
I've refactored the method to run on a single worker; this makes the API more homogenous with _delete_worker_data too.